diff --git a/src/test/authentication/t/004_profile.pl b/src/test/authentication/t/004_profile.pl
index b01311b4f04..581fd678882 100644
--- a/src/test/authentication/t/004_profile.pl
+++ b/src/test/authentication/t/004_profile.pl
@@ -14,7 +14,7 @@ $node->start;
 $node->safe_psql('postgres', 'create user profile_admin superuser;');
 # We want to access our database with superuser role without any passwords.
 unlink($node->data_dir . '/pg_hba.conf');
-if ($Config{osname} eq 'MSWin32') {
+if ($windows_os) {
 	$node->append_conf('pg_hba.conf', "host all profile_admin 127.0.0.1/32 trust");
 	$node->append_conf('pg_hba.conf', "host all all 127.0.0.1/32 password");
 } else {
diff --git a/src/test/authentication/t/005_password_policies_on_standby.pl b/src/test/authentication/t/005_password_policies_on_standby.pl
index ed89df97b03..498ea2db339 100644
--- a/src/test/authentication/t/005_password_policies_on_standby.pl
+++ b/src/test/authentication/t/005_password_policies_on_standby.pl
@@ -192,7 +192,7 @@ $node_primary->init(allows_streaming => 1);
 # 1. Configure connections so that we use password policies almost every time.
 
 unlink($node_primary->data_dir . '/pg_hba.conf');
-if ($Config{osname} eq 'MSWin32')
+if ($windows_os)
 {
 	$node_primary->append_conf(
 		'pg_hba.conf', q(
